microsoft/qdk-python
# Custom credential types for better user experience

This PRs improves the customer experience when authenticating with the Azure Quantum SDK for Python while still supporting track2 generated clients and more modern Azure.Identity credentials.

Users can pass any compatible credential with Azure.Identity, but if none is passed we will use a default custom credential based on the DefaultAzureCredential.
The custom code have the following benefits:
1) It still tries multiple authentication methods like the DefaultAzureCredential, but it will not print warning messages if an authentication method has failed but someone other succeeded. The warning messages are confusing to the user because they were emitted even if the final authentication succeeded. This mitigates issue [Python SDK 18972](https://github.com/Azure/azure-sdk-for-python/issues/18972)
2) It will not halt the other authentication methods if one of them failed. It will continue to try all the other authentication methods. This help mitigate issues with when VS Credential have expired (this is by-design according to the Azure SDK team).
3) It will automatically detect and set the `tenant_id` for the credentials that need  it (for example: `InteractiveBrowserCredential`) as a workaround for bug [Python SDK #18975](https://github.com/Azure/azure-sdk-for-python/issues/18975), which affects users with MSA/personal accounts.

### How to test it:

Key scenarios to be tested:
1. Interactive authentication with work account (your @microsoft.com account)
2. Interactive authentication with personal account (your @live.com/outlook.com account)
3. Automatic authentication via VisualStudio Code (you need to sign-in in your VS Code)
4. Automatic authentication via Az CLI (you need to execute az login )
5. Automated authentication via Environment Credentials

The SDK will try multiple authentication methods, in this order:
1. **EnvironmentCredential**
2. ManagedIdentityCredential
3. SharedTokenCacheCredential
4. **VisualStudioCodeCredential**
5. **AzureCliCredential**
6. AzurePowerShellCredential
7. **InteractiveBrowserCredential**

So to test the **InteractiveBrowserCredential** the previous credentials should fail (i.e. you need to sign-out from VS Code and execute az logout )
 
Steps to test:
1. Checkout the PR branch, go to `azure-quantum` folder and run `pip install -e .`
2. logout from VS Code (in case you have logged-in before)
3. run `az logout`
4. Follow the steps from our [quick-start](https://docs.microsoft.com/en-us/azure/quantum/quickstart-microsoft-qio#create-a-workspace-object-in-your-python-code-and-log-in). Or simply use the template below: 
```python
from azure.quantum import Workspace

workspace = Workspace(
    resource_id="/subscriptions/[subscription_id]/resourceGroups/[resource_group_name]/providers/Microsoft.Quantum/Workspaces/[workspace_name]",
    location="[location]"
)

from azure.quantum.optimization import Problem, ProblemType, Term
from azure.quantum.optimization import ParallelTempering

solver = ParallelTempering(workspace, timeout=100) # timeout in seconds
terms = [Term(c=1, indices=[0])]
problem = Problem(name="My Problem123", problem_type=ProblemType.ising, terms=terms)
result = solver.optimize(problem)
solutions = result.get('solutions')
expected_solution = {'configuration': {'0': -1}, 'cost': -1.0}
assert solutions[0] == expected_solution

print("Success" if solutions[0] == expected_solution else "Failure")
``` 

When submitting the job, it should open a browser page for you to login.
First try to login with a work account, then reset the kernel and try to login with a personal account (you will need to adjust the resource_id to connect to the workspace in the two different accounts).
 
Then you can try other auth methods:
1) Run az login , reset the kernel, and then it will use that credential first
2) Login on VS Code, reset the kernel, and then it will use that credential first
3) Set the folowing environment variables to use the EnvironmentCredential first
    - AZURE_CLIENT_ID
    - AZURE_CLIENT_SECRET
    - AZURE_TENANT_ID
